cmd/prometheus/main.go

```diff
@@ -272,42 +267,22 @@ func main() {
 	}
 
 	// Depends on cfg.web.ScrapeManager so needs to be after cfg.web.ScrapeManager = scrapeManager
-	webHandler := web.New(log.With(logger, "component", "web"), &cfg.web)
+	webHandler := web.New(log.With(logger, "component", "web"), &cfg.web, configManager)
 
 	// Monitor outgoing connections on default transport with conntrack.
 	http.DefaultTransport.(*http.Transport).DialContext = conntrack.NewDialContextFunc(
 		conntrack.DialWithTracing(),
 	)
 
-	reloaders := []func(cfg *config.Config) error{
-		remoteStorage.ApplyConfig,
-		webHandler.ApplyConfig,
-		notifier.ApplyConfig,
-		discoveryManager.ApplyConfig,
-		scrapeManager.ApplyConfig,
-		func(cfg *config.Config) error {
-			// Get all rule files matching the configuration oaths.
-			var files []string
-			for _, pat := range cfg.RuleFiles {
-				fs, err := filepath.Glob(pat)
-				if err != nil {
-					// The only error can be a bad pattern.
-					return fmt.Errorf("error retrieving rule files for %s: %s", pat, err)
-				}
-				files = append(files, fs...)
-			}
-			return ruleManager.Update(time.Duration(cfg.GlobalConfig.EvaluationInterval), files)
-		},
+	reloaders := []func(cfg config.ReloadReader) error{
+		remoteStorage.Reload,
+		notifierManager.Reload,
+		discoveryManager.Reload,
+		ruleManager.Reload,
 	}
```

Comment:
My main concern here would be that all packages keep relying on the `config` package. Would it be reasonable to just give them all `Reload` method but with different signatures so each only gets passed what it needs. In the `main` package one could still reduce this down to a unified `reload` method as we did in the removed code for the rule manager?

I know it makes main a bit verbose but for sure decouples things a bit.

Reply:
just trying something close to what you suggest , but the entanglement is a lot deeper than the rule manager!

https://github.com/prometheus/prometheus/pull/3629 will wait for this PR and will continue the work.